server/client.go

```diff
@@ -301,73 +365,202 @@ func (c *client) readLoop() {
+// Lock must be held
+func (c *client) flushOutbound() bool {
+	if c.flags.isSet(flushOutbound) {
+		return false
+	}
+	c.flags.set(flushOutbound)
+	defer c.flags.clear(flushOutbound)
+
+	// Check for nothing to do.
+	if c.nc == nil || c.srv == nil || c.out.pb == 0 {
+		return true // true because no need to queue a signal.
+	}
+
+	// Snapshot opts
+	srv := c.srv
+
+	// Place primary on nb, assign primary to secondary, nil out nb and secondary.
+	nb := c.collapsePtoNB()
+	c.out.p, c.out.nb, c.out.s = c.out.s, nil, nil
+
+	// For selecting primary replacement.
+	cnb := nb
+
+	// In case it goes away after releasing the lock.
+	nc := c.nc
+	attempted := c.out.pb
+	apm := c.out.pm
+
+	// Do NOT hold lock during actual IO
+	c.mu.Unlock()
+
+	// flush here
+	now := time.Now()
+	// FIXME(dlc) - writev will do multiple IOs past 1024 on
+	// most platforms, need to account for that with deadline?
+	nc.SetWriteDeadline(now.Add(c.out.wdl))
+	// Actual write to the socket.
+	n, err := nb.WriteTo(nc)
+	nc.SetWriteDeadline(time.Time{})
+	lft := time.Since(now)
+
+	// Re-acquire client lock
+	c.mu.Lock()
+
+	// Update flush time statistics
+	c.out.lft = lft
+
+	// Subtract from pending bytes and messages.
+	c.out.pb -= n
+	c.out.pm -= apm // FIXME(dlc) - this will not be accurate.
+
+	// Check for partial writes
+	if n != attempted && n > 0 {
+		c.handlePartialWrite(nb)
+	} else if n >= int64(c.out.sz) {
+		c.out.sws = 0
+	}
+
+	if err != nil {
+		if n == 0 {
+			c.out.pb -= attempted
+		}
+		c.clearConnection()
+		if ne, ok := err.(net.Error); ok && ne.Timeout() {
+			atomic.AddInt64(&srv.slowConsumers, 1)
+			c.Noticef("Slow Consumer Detected: WriteDeadline of %v Exceeded", c.out.wdl)
+		} else {
+			c.Debugf("Error flushing: %v", err)
+		}
+		return true
+	}
 
-		// Update buffer size as/if needed.
+	// Adjust based on what we wrote plus any pending.
+	pt := int(n + c.out.pb)
 
-		// Grow
-		if n == len(b) && len(b) < maxBufSize {
-			b = make([]byte, len(b)*2)
+	// Adjust sz as needed downward, keeping power of 2.
+	// We do this at a slower rate, hence the pt*4.
+	if pt < c.out.sz && c.out.sz > minBufSize {
+		c.out.sws++
+		if c.out.sws > shortsToShrink {
+			c.out.sz >>= 1
 		}
+	}
+	// Adjust sz as needed upward, keeping power of 2.
+	if pt > c.out.sz && c.out.sz < maxBufSize {
+		c.out.sz <<= 1
+	}
 
-		// Shrink, for now don't accelerate, ping/pong will eventually sort it out.
-		if n < len(b)/2 && len(b) > minBufSize {
-			b = make([]byte, len(b)/2)
+	// Check to see if we can reuse buffers.
+	if len(cnb) > 0 {
+		oldp := cnb[0][:0]
+		if cap(oldp) >= c.out.sz {
+			// Replace primary or secondary if they are nil, reusing same buffer.
+			if c.out.p == nil {
+				c.out.p = oldp
+			} else if c.out.s == nil || cap(c.out.s) < c.out.sz {
+				c.out.s = oldp
+			}
 		}
 	}
+	return true
+}
+
+// flushSignal will use server to queue the flush IO operation to a pool of flushers.
+// Lock must be held.
+func (c *client) flushSignal() {
+	c.out.sg.Signal()
```

Comment:
Should we have an `out.inWait = true` and `out.inWait = false` surrounding the call to `sg.Wait()` and flush only if `inWait` is true? That would possibly reduce the number of `Signal()` calls.

Reply:
So far has not been a problem. Also, when Wait returns it will have the lock but it is not guaranteed the next instruction to reset the flag gets run.